pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py

```diff
@@ -244,11 +244,10 @@ def activate_ordering_keys(self, ordering_keys):
         ordering keys, if they exist. Clean up state for keys that no longer
         have any queued messages.
 
-        There is enough load capacity to send one message per ordering key
-        because each key is the result of an ack or nack. Since the load went
-        down by one message, it's safe to send the user another message for the
-        same key. This is why this method does not check the load before
-        activating the key.
+        Since the load went down by one message, it's probably safe to send the
+        user another message for the same key. Since the released message may be
+        bigger than the previous one, this may increase the load above the maximum.
```

Comment:
Just to double check - it might also happen that we release a message while the current load is _still above 1.0_. Please confirm that this is fine, too, if it happens due to message size.

---

**An example how this could happen** (please correct me if I'm wrong):

- `FlowControl.max_messages` is set to "a lot" and `FlowControl.max_bytes` is set to 10. The user code is currently processing three messages with sizes 1, 1, and 1, respectively (load == 0.3).
- The client then receives several message with sizes 9, 8, and 7. Since the current load allows for it, we dispatch the first message (size 9). The user code is now processing messages of sizes 1, 1, 1, and 9, the load is 1.2, and two messages with sizes 8 and 7 are on hold.
- The user code ACKs a small message, that message is dropped from the lease management. The load drops to 1.1.
- Since the load went down by one message, we (incorrectly) assume that we now have capacity to release another message, thus we release the message of size 8.
- The user code is now processing messages of sizes 1, 1, 9, and 8, and the load actually jumps to 1.9.
- In unlucky cases, the issue can compound - the user code ACKs another small message (size 1), which makes the client to release another (bigger) message - size 7 this time. The user code is then processing messages of sizes 1, 9, 8, 7, and the load is 2.5 ...

Reply:
Yes, you're right that this can happen.

My assumption is that messages are of roughly the same size and that user byte size limits are significantly higher than the messages themselves. The default byte size limit is 100mb and the max user-message size is 10mb, which I believe is rare.

Hopefully, users wont hit upon a pathological case often.